modules/file-io/src/test/java/org/apache/ignite/internal/fileio/AbstractFileIoTest.java

```diff
@@ -0,0 +1,588 @@
+
+        assertEquals(1024, fileIo.writeFully(ByteBuffer.wrap(copyOfRange(randomBytes, 2 * 1024, 3 * 1024))));
+        assertEquals(3 * 1024, fileIo.position());
+
+        assertEquals(1024, fileIo.writeFully(ByteBuffer.wrap(copyOfRange(randomBytes, 3 * 1024, 4 * 1024))));
+        assertEquals(4 * 1024, fileIo.position());
+
+        assertEquals(0, fileIo.writeFully(ByteBuffer.wrap(new byte[0])));
+        assertEquals(4 * 1024, fileIo.position());
+
+        fileIo.force();
+
+        assertArrayEquals(randomBytes, toByteArray(testFilePath));
+    }
+
+    @Test
+    void testWriteByPositionFully() throws Exception {
+        byte[] randomBytes = randomByteArray(4 * 1024);
+
+        Path testFilePath = workDir.resolve("test");
+
+        FileIo fileIo = fileIoFactory.create(testFilePath);
+
+        assertEquals(1024, fileIo.writeFully(ByteBuffer.wrap(copyOfRange(randomBytes, 1024, 2 * 1024)), 1024));
+        assertEquals(0, fileIo.position());
+
+        assertEquals(1024, fileIo.writeFully(ByteBuffer.wrap(copyOfRange(randomBytes, 3 * 1024, 4 * 1024)), 2 * 1024));
+        assertEquals(0, fileIo.position());
+
+        assertEquals(0, fileIo.writeFully(ByteBuffer.wrap(new byte[0]), 3 * 1024));
+        assertEquals(0, fileIo.position());
+
+        fileIo.force();
+
+        byte[] expectedBytes = new byte[3 * 1024];
+
+        System.arraycopy(randomBytes, 1024, expectedBytes, 1024, 1024);
+        System.arraycopy(randomBytes, 3 * 1024, expectedBytes, 2 * 1024, 1024);
+
+        assertArrayEquals(expectedBytes, toByteArray(testFilePath));
+    }
+
+    @Test
+    void testWriteFromByteArrayFully() throws Exception {
+        byte[] randomBytes = randomByteArray(4 * 1024);
+
+        Path testFilePath = workDir.resolve("test");
+
+        FileIo fileIo = fileIoFactory.create(testFilePath);
+
+        assertEquals(1024, fileIo.writeFully(randomBytes, 0, 1024));
+        assertEquals(1024, fileIo.position());
+
+        assertEquals(1024, fileIo.writeFully(randomBytes, 1024, 1024));
+        assertEquals(2 * 1024, fileIo.position());
+
+        assertEquals(1024, fileIo.writeFully(randomBytes, 2 * 1024, 1024));
+        assertEquals(3 * 1024, fileIo.position());
+
+        assertEquals(1024, fileIo.writeFully(randomBytes, 3 * 1024, 1024));
+        assertEquals(4 * 1024, fileIo.position());
+
+        assertEquals(0, fileIo.writeFully(randomBytes, 4 * 1024, 0));
+        assertEquals(4 * 1024, fileIo.position());
+
+        fileIo.force();
+
+        assertArrayEquals(randomBytes, toByteArray(testFilePath));
+    }
+
+    @Test
+    void testMap() throws Exception {
+        Path testFilePath = workDir.resolve("test");
+
+        FileIo fileIo = fileIoFactory.create(testFilePath);
+
+        MappedByteBuffer mmap = fileIo.map(1024);
+
+        assertNotNull(mmap);
+
+        assertEquals(0, mmap.position());
+
+        assertEquals(1024, mmap.capacity());
+    }
+
+    @Test
+    void testForce() throws Exception {
```

Comment:
This one is kinda pointless, but let's leave it. How do you even test that sync has actually been invoked? I've no idea!

Reply:
Here I check only the basic functionality: that there will be no errors when calling **FileIo#force** and that the data will actually be flushed to disk.

Reply:
Added a comment.